src/hotspot/share/opto/escape.cpp

```diff
@@ -401,6 +416,239 @@ bool ConnectionGraph::compute_escape() {
   return has_non_escaping_obj;
 }
 
+// Check if it's profitable to reduce the Phi passed as parameter.  Returns true
+// if at least one scalar replaceable allocation participates in the merge and
+// no input to the Phi is nullable.
+bool ConnectionGraph::can_reduce_this_phi_inputs(PhiNode* phi) const {
+  // Check if there is an scalar replaceable allocate in the Phi
+  bool found_sr_allocate = false;
+
+  for (uint i = 1; i < phi->req(); i++) {
+    // Right now we can't restore a "null" pointer during deoptimization
+    if (_igvn->type(phi->in(i))->maybe_null()) {
+      NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Can NOT reduce Phi %d. Input %d is nullable.", phi->_idx, i);)
+      return false;
+    }
+
+    // We are looking for at least one SR object in the merge
+    JavaObjectNode* ptn = unique_java_object(phi->in(i));
+    if (ptn != NULL && ptn->scalar_replaceable()) {
+      assert(ptn->ideal_node() != NULL && ptn->ideal_node()->is_Allocate(), "sanity");
+      Node* alloc = ptn->ideal_node()->as_Allocate();
+
+      if (PhaseMacroExpand::can_eliminate_allocation(_igvn, alloc->as_Allocate(), NULL, true)) {
+        found_sr_allocate = true;
+      } else {
+        ptn->set_scalar_replaceable(false);
+      }
+    }
+  }
+
+  NOT_PRODUCT(if (TraceReduceAllocationMerges && !found_sr_allocate) tty->print_cr("Can NOT reduce Phi %d. No SR Allocate as input.", phi->_idx);)
+  return found_sr_allocate;
+}
+
+// Check if we are able to untangle the merge. Right now we only reduce Phis
+// which are only used as debug information.
+bool ConnectionGraph::can_reduce_this_phi_users(PhiNode* phi) const {
+  for (DUIterator_Fast imax, i = phi->fast_outs(imax); i < imax; i++) {
+    Node* use = phi->fast_out(i);
+
+    if (use->is_SafePoint()) {
+      if (use->is_Call() && use->as_Call()->has_non_debug_use(phi)) {
+        NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Can NOT reduce Phi %d. Call has non_debug_use().", phi->_idx);)
+        return false;
+      }
+    } else {
+      NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Can NOT reduce Phi %d. One of the uses is: %d %s", phi->_idx, use->_idx, use->Name());)
+      return false;
+    }
+  }
+
+  return true;
+}
+
+// Returns true if: 1) It's profitable to reduce the merge, and 2) It The Phi is
+// only used in some certain code shapes. Check comments in
+// 'can_reduce_this_phi_inputs' and 'can_reduce_this_phi_users' for more
+// details.
+bool ConnectionGraph::can_reduce_this_phi(PointsToNode* var) const {
+  assert(var->escape_state() == PointsToNode::NoEscape, "Phi should be NoEscape");
+
+  PhiNode* phi = var->ideal_node()->as_Phi();
+  const Type* phi_t = _igvn->type(phi);
+  if (phi_t == NULL || phi_t->isa_instptr() == NULL || !phi_t->is_instptr()->klass_is_exact()) {
+    return false;
+  }
+
+  if (!can_reduce_this_phi_inputs(phi) || !can_reduce_this_phi_users(phi)) {
```

Comment:
I'm wondering whether there are better names for these methods. They kind of read as if we're checking whether we can reduce Phi inputs or reduce Phi users (as opposed to reducing a Phi), but really, we're checking inputs and users. Maybe `can_reduce_this_phi_check_inputs` and `can_reduce_this_phi_check_users`? Or maybe put all the functionality into this function?

Reply:
I think the names you suggested read definitely better. Thanks.